### DISCUSSION TITLE
Role based authorization in Next.js middleware - not possible to select rows in Supabase

### DISCUSSION THREAD

**Discussion by riccardolardi** [supabase/supabase#29482] [Questions]
I've set up Supabase in Next.js middleware according to the [docs](https://supabase.com/docs/guides/auth/server-side/creating-a-client?queryGroups=framework&framework=nextjs&queryGroups=environment&environment=middleware) 

Now what I would like to do is protect some routes depending on a row (`role`) of a table (`profiles`) which I would need to query on each request. This doesn't seem to work, when doing the select using supabase server client I always get an empty row:

`profile: { error: null, data: [], count: null, status: 200, statusText: 'OK' }`

The table is secured by RLS but even using the service role key does not result in any data.

Here's my implementation of `utils/supabase/middleware.ts`

```ts
import { env } from "-/src/env";
import { createServerClient } from "@supabase/ssr";
import { NextResponse, type NextRequest } from "next/server";

export async function updateSession(
  request: NextRequest,
  response: NextResponse,
) {
  const supabase = createServerClient(
    env.NEXT_PUBLIC_SUPABASE_URL,
    env.NEXT_PUBLIC_SUPABASE_ANON_KEY // also tried using the service role key, to no avail
    {
      cookies: {
        getAll() {
          return request.cookies.getAll();
        },
        setAll(cookiesToSet) {
          response = NextResponse.next({ request });
          cookiesToSet.forEach(({ name, value, options }) => {
            request.cookies.set(name, value);
            response.cookies.set(name, value, options);
          });
        },
      },
    },
  );

  const {
    data: { user },
  } = await supabase.auth.getUser();
  const profile = await supabase
    .from("profiles")
    .select("*")
    .eq("id", user?.id)
    .single();
  console.log({ profile }); // { profile: { error: null, data: [], count: null, status: 200, statusText: 'OK' } }
}
```

Is it not possible to make queries using supabase client from within the Next.js middleware?

Thanks for any help.

**Comment by bytaesu** [comment]
Using getUser(), basic setup is possible. However, middleware should remain lightweight. 

The official Next.js documentation here. Even if the task is ‘theoretically’ possible, according to the Next.js documentation, data processing should be done separately.

https://nextjs.org/docs/app/building-your-application/routing/middleware#use-cases

![image](https://github.com/user-attachments/assets/ee3f544c-5103-4068-aeb3-248d6b11aeab)

**Reply by riccardolardi** [reply]
Thanks @bytaesu 

Can you recommend a way to check for the role instead? To my knowledge, it is not completely safe to rely on layout based authorization since it can be bypassed in certain ways.

**Reply by bytaesu** [reply]
### Middleware
Use middleware with getUser() to perform a basic check such as if the user is logged in

### Server-Component
In here, create supabase client using @/supabase/ssr
Conduct additional operations to verify the user

### Client-Component
Build UI components that are related to user information.

### +) Check this repository

> https://github.com/midday-ai/midday/blob/main/apps/dashboard/src/app/%5Blocale%5D/(app)/setup/page.tsx

There’s an open-source project called Midday, and it uses Supabase. It’s written with App Router, and I remember seeing user verification done in the way I mentioned above. I’m not an expert either, but referencing open-source projects can be really helpful!
```
export const getTeamUser = async () => {
  const supabase = createClient();
  const { data } = await getUser();

  return unstable_cache(
    async () => {
      return getTeamUserQuery(supabase, {
        userId: data.id,
        teamId: data.team_id,
      });
    },
    ["team", "user", data.id],
    {
      tags: [`team_user_${data.id}`],
      revalidate: 180,
    },
  )(data.id);
};

```

**Reply by riccardolardi** [reply]
Thanks a lot. I would think though there would be another, more centralized / HOC option to run authorization checks on a request basis rather than doing them in every single page.tsx, no?

**Reply by bytaesu** [reply]
This is the method I’m thinking of.

Make custom hooks and use it on layout.tsx (AppRouter)

or

Make container such as AuthWrapper/AuthProvider

**Reply by riccardolardi** [reply]
Hooks would only run on client side - I would much rather do those checks as early as possible, if not in middleware than at render level

**Reply by bytaesu** [reply]
You can execute the query on the server side. Something like a server action.

https://github.com/midday-ai/midday/blob/main/packages/supabase/src/queries/cached-queries.ts#L67

**Reply by bytaesu** [reply]
AND,

This is supabase's approach.
While it's not about utilizing Supabase but rather Supabase itself, the concept can still be explored in a similar way.

https://github.com/supabase/supabase/blob/master/apps/studio/data/organizations/organization-roles-query.ts


I don't think there's a definitive answer. Some people might prefer to use client-side rendering only.
(Not me 🤓)

**Reply by riccardolardi** [reply]
Thanks @bytaesu I will take an approach where I will create a `protectRoute()` server action and make it configurable. And then use it in all page.tsx files. Still it would be nice to be able to have a slightly complex (not middleware) "generic" check higher up the tree covering an entire part of the app, not needing to repeat in every single page.tsx...

**Reply by bytaesu** [reply]
(In case of AppRouter)

You can perform that in layout.tsx. It wraps all the child page.tsx files.

I understand what you’re thinking about. You want to reduce repetitive tasks and write ‘well-designed’ code. I usually think about that too.

Most of the answers are in the official documentation, and some others can be found in open-source projects.

Good luck!
